### PR TITLE
FIX-#4636: allows `read_parquet` to detect column partitioning in non-local filesystems

### LINKED ISSUE #4636: BUG: read_parquet can't detect column partitioning in non-local filesystems
### System information
- **Modin version** (`modin.__version__`): 86d361078fbad03143bf02ce212ffa8af4d416b5

Parquet datasets are arbitrarily deep directories of data files that can include column partitioning (ref. [here](https://spark.apache.org/docs/2.4.0/sql-data-sources-parquet.html#partition-discovery)). In column partitioning, a directory like `col1=3` only has data where column 1 is equal to 3. Modin tries to default to pandas for datasets with column partitioning, but it [assumes](https://github.com/modin-project/modin/blob/5ca6d40f0ab9e6fede0dd5e77a415231a394fa90/modin/core/io/column_stores/parquet_dispatcher.py#L59) that column partitioning can only happen when the path to read is a **local** directory. The directory could, for example, be on s3.

I think that nothing is functionally wrong when we fail to detect column partitioning. Modin can still read the data correctly. It's just not defaulting to pandas even though we expect it to.

### PR DESCRIPTION
**Pull request by Billy2551**
Signed-off-by: Bill Wang <billiam@ponder.io>

## What do these changes do?

Added support for `read_parquet` on column partitioned parquet files in non-local filesystems. Includes tests on non-local S3 parquet files.

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4636 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/io/column_stores/parquet_dispatcher.py [hunk 1]
```diff
@@ -609,18 +609,24 @@ def _read(cls, path, engine, columns, **kwargs):
         )
         from modin.pandas.io import PQ_INDEX_REGEX
 
-        if isinstance(path, str) and os.path.isdir(path):
+        if isinstance(path, str):
+            if os.path.isdir(path):
+                path_generator = os.walk(path)
+            else:
+                storage_options = kwargs.get("storage_options")
+                fs, fs_path = url_to_fs(path, **storage_options)
```
Comment (anmyachev) [inline modin/core/io/column_stores/parquet_dispatcher.py:617]: `storage_options` variable can be `None`, so this construction breaks several tests in CI.

File: modin/core/io/column_stores/parquet_dispatcher.py [hunk 2]
```diff
@@ -609,18 +609,27 @@ def _read(cls, path, engine, columns, **kwargs):
...
             # Thus, we can use os.walk(), which does a dfs search, to walk
             # through the different columns that the data is partitioned on
-            for (_, dir_names, files) in os.walk(path):
+            for (_, dir_names, files) in path_generator:
                 if dir_names:
                     partitioned_columns.add(dir_names[0].split("=")[0])
                 if files:
                     # Metadata files, git files, .DSStore
-                    if files[0][0] == ".":
+                    if len(files[0]) > 0 and files[0][0] == ".":
```
Comment (mvashishtha) [inline modin/core/io/column_stores/parquet_dispatcher.py:632]: Do you know why we only check the first file in the directory? i.e. `files[0]`?

File: modin/core/io/column_stores/parquet_dispatcher.py (hunk 1 above)
Comment (Billy2551) [reply modin/core/io/column_stores/parquet_dispatcher.py]: Done.

File: modin/core/io/column_stores/parquet_dispatcher.py (hunk 2 above)
Comment (mvashishtha) [reply modin/core/io/column_stores/parquet_dispatcher.py:632]: If I understand the current logic correctly, it’s “if there’s any directory within path where the first file [and apparently the files don’t come in any guaranteed order] starts with ‘.’, then stop looking for partitioned columns”. I think is buggy in a few ways, e.g. see #4637. For now could you just add a TODO to fix the logic here?

**Review by pyrito — APPROVED** [review]

**Review by mvashishtha — APPROVED** [review]
Thanks @Billy2551 for the fix!

**Pull request merged**